Update KEP-2076: Kueuectl. (#3101)
kubernetes-sigs/kueue: keps/2076-kueuectl/README.md

Output:
None.

> ### Delete Workload

Reviewer: Adding context and motivations for dedicated commands would be helpful to understand this command objective.

Author: Done

* `describe workload|wl|clusterqueue|cq|localqueue|lq|resourceflavor|rf`
* `edit workload|wl|clusterqueue|cq|localqueue|lq|resourceflavor|rf`
* `patch workload|wl|clusterqueue|cq|localqueue|lq|resourceflavor|rf`
* `delete clusterqueue|cq|localqueue|lq|resourceflavor|rf`

### Delete Workload

Motivation:
The pass-through delete workload command doesn't actually delete the Workload but 
instead requeues it. This can confuse users, as the Workload remains listed even 
after they attempt to delete it, due to its recreation.

> This command will delete the Workload and its corresponding Job(s). If the Workload 

Reviewer 2: Does it need to delete the workload object? I think deleting the Job is enough - the workload object will be deleted by Kueue by this code:

Reviewer 2: Ok, but this is a corner case. If the Job exists, no point deleting the workload.

Reviewer 2: hm, or maybe there is a point - to make the operation synchronous - so that the workload is gone if the `kueuectl workload list` command is repeated very fast. If this is the motivation, then please state it in the KEP, otherwise we can just say in the help that the operation is async and may take a while.

Reviewer 3: (prebuild workloads are orphan in multikueue worker clusters until they get admitted)

Author: Done

> This command will delete the Workload and its corresponding Job(s). If the Workload 

Reviewer 3: I think we should better define what "corresponding Job(s)" are. 
And maybe think about:
- how we ca address what I mentioned in https://github.com/kubernetes-sigs/kueue/pull/2992#discussion_r1765051596
- maybe mention the the effect of deleting parent object on workload sibling objects

Author: > how we ca address what I mentioned in https://github.com/kubernetes-sigs/kueue/pull/2992#discussion_r1765051596

We just looked at this case with @mimowo and found that the Deployment recreates the Pod and it will have a different name.

Reviewer 3: I don't understand. The Deployment is just an example.

Author: The Workload associated with pod and we delete this Pod. So it's ok here. But I also need to check [FluxMiniclusters](https://kueue.sigs.k8s.io/docs/tasks/run/flux_miniclusters/).

Author: A user may want to delete the whole `Deployement`, but for that, we need to have a different workload.

Reviewer 3: Even if the names are slightly different in case of replicasets/deployments, you end up deleting two objects and other two (having the same purpose) are created.

Reviewer 2: I think this is ok for now, as the workload requested to be deleted by the user is gone, which was the main confusing factor in the original issue.

The support for Deployments is in initial phase.

Long term, we discuss introducing a concept of a workload fragment. Then workloads for Pods will be considered workload fragments, and we may display a parent workload which aggregates the fragments. Then deleting the parent workload would delete the Deployment instance, but I believe this work can be deferred for later, but cc @mwielgus in case we want to prioritize it more as part of the fix.

Author: > But I also need to check [FluxMiniclusters](https://kueue.sigs.k8s.io/docs/tasks/run/flux_miniclusters/).

Just to confirm, the situation in `FluxMiniclusters` is the same. It removes the Job associated with a Workload but then recreates both a new Job and a new Workload.

Reviewer 3: > Just to confirm, the situation in `FluxMiniclusters` is the same. It removes the Job associated with a Workload but then recreates both a new Job and a new Workload.

With the same name ore a slightly different one?

Author: With different one.

owns any Jobs, the command will prompt for deletion approval, displaying the Jobs 
that will be deleted along with the Workload.

Format:
```

> kueuectl delete workload name

Reviewer 2: how a user passes the namespace, `-n`?

Author: The `-n` and `--namespaces`  flags is exist by default on for each command. But I will add it.

Author: Done

```

Flags:
```

> -y, --yes             Confirm the deletion of the Workload and its corresponding Job.

Reviewer 2: ```suggestion
-y, --yes             Automatic yes to the prompt for deleting the Workload.
```

Author: Fixed

    --all             Delete all Workloads, in the specified namespace.
    --all-namespaces  Using with --all flag. Delete all Workloads, in all namespaces.
```

Approval Output:
```
This operation will also delete:

>   - jobs.batch/name associated with the name workload

Reviewer 2: ```suggestion
  - jobs.batch/name associated with the namespace/name workload
```
?

Author: Done

Do you want to proceed (y/n)?
```

Deletion Process Output:
```

> workload.kueue.x-k8s.io/name deleted

Reviewer 2: I think listing Job object is more important. We may additionally display the workload object name too.

* `describe workload|wl|clusterqueue|cq|localqueue|lq|resourceflavor|rf`
* `edit workload|wl|clusterqueue|cq|localqueue|lq|resourceflavor|rf`
* `patch workload|wl|clusterqueue|cq|localqueue|lq|resourceflavor|rf`
* `delete clusterqueue|cq|localqueue|lq|resourceflavor|rf`

### Delete Workload

Motivation:

> The pass-through delete workload command doesn't actually delete the Workload but 
> instead requeues it. This can confuse users, as the Workload remains listed even 
> after they attempt to delete it, due to its recreation.

Reviewer 3: ```suggestion
The pass-through delete workload command deletes the Workload API object, which is later
recreated leading to its job to be re-queued. This can confuse users, as the recreated Workload
has the same name giving the impression that the `delete workload` command did not succeed.
```

Author: Fixed
